pkg/gitparse/gitparse.go

```diff
@@ -614,6 +627,47 @@ func isCommitLine(isStaged bool, latestState ParseState, line []byte) bool {
 	return false
 }
 
+func parseCommitLine(line []byte) (hash []byte, ref []byte) {
+	// Check that the commit line contains a 40-character hash and set it.
+	// `commit e5575cd6f2d21d3a1a604287c7bf4a7eab2266e0\n`
+	if len(line) >= 47 {
+		hash = line[7:47]
+	}
+
+	// Check if the commit line includes branch references.
+	// `commit 2dbbb28727c7c2954438666dafba57bb8c714d3b refs/heads/fix/github-enterprise-gist\n`
+	if len(line) > 48 {
+		ref = line[48 : len(line)-1]
+	}
+
+	return
+}
+
+// ParseCommitSource s
+// https://git-scm.com/docs/git-log#Documentation/git-log.txt---source
+func parseSourceRef(ref []byte) string {
+	// We don't care about 'normal' refs.
+	if bytes.HasPrefix(ref, []byte("refs/heads/")) || bytes.HasPrefix(ref, []byte("refs/tags/")) {
+		return ""
+	}
+
+	// Handle GitHub pull requests.
+	// e.g., `refs/pull/238/head` or `refs/pull/1234/merge`
+	if after, ok := bytes.CutPrefix(ref, []byte("refs/pull/")); ok {
+		prNumber := after[:bytes.Index(after, []byte("/"))]
+		return "Pull request #" + string(prNumber)
+	}
+
+	// Handle GitLab merge requests
+	// e.g., `refs/merge-requests/238/head` or `refs/merge-requests/1234/merge`
+	if after, ok := bytes.CutPrefix(ref, []byte("refs/merge-requests/")); ok {
+		mrNumber := after[:bytes.Index(after, []byte("/"))]
+		return "Merge request #" + string(mrNumber)
+	}
+
+	return fmt.Sprintf("%s (hidden ref)", string(ref))
```

Comment:
Does "hidden" just mean "not visible via the respective web UI?" My small concern here is that this also happens if we simply fail to parse the relevant ref because we made a mistake somewhere. What do you think of just returning the ref as-is in that case - would it be confusing?